python/ray/data/grouped_dataset.py

```diff
@@ -144,6 +146,72 @@ def _aggregate_on(
         )
         return self.aggregate(*aggs)
 
+    def map_groups(
+        self,
+        fn: Union[CallableClass, Callable[[BatchType], BatchType]],
+        *,
+        compute: Optional[str] = None,
+        batch_format: str = "native",
+        **ray_remote_args
+    ) -> "Dataset[Any]":
+        """Apply a function to each group of records.
+
+        The user given function should take as argument a batch, which contains
+        exactly the records for a group, and return a block of the same batch type.
+        The results of different groups will be collected to form the result dataset.
+        """
+
+        # Globally sort records by key.
+        # Note that sort() will ensure that records of the same key partitioned
+        # into the same block.
+        sorted_ds = self._dataset.sort(self._key)
```

Comment:
self._key can be None which means the entire dataset is one group, however None has a different meaning for sort(). We should skip sorting if group key is None (lets also add a test case for this as well).

Reply:
Done. Thanks for pointing this out!
On the other hand, I think this disconnection makes it not work for groupby(None): we are relying on sort() to make same keys in same block; if sort() is not going to do that for None (assuming we can put all rows in one node in terms of resource),  then map_groups() will not work correctly.